File: frontend/src/lib/emailVariables.ts
/**
 * Values that the server puts in an email when it sends. Keep the keys and fallbacks in step with
 * VARIABLES in bwh_os/mailing/email_variables.py.
 */
export interface EmailVariable {
	key: string
	label: string
	description: string
	/** Used when the value is empty */
	fallback?: string
	/** Shown in the preview */
	sample: string
}

/** What to type for this variable, which is not its label. */
export function token(variable: EmailVariable): string {
	return `{{ ${variable.key} }}`
}

/** The tooltip text of a variable. It names the token, because a link field has no chips. */
export function describe(variable: EmailVariable): string {
	const fallback = variable.fallback ? ` When empty: “${variable.fallback}”.` : ''
	return `${token(variable)} — ${variable.description}${fallback}`
}

const FIRST_NAME: EmailVariable = {
	key: 'first_name',
	label: 'First name',
	description: "The subscriber's first name.",
	fallback: 'there',
	sample: 'Priya',
}

const EMAIL: EmailVariable = {
	key: 'email',
	label: 'Email',
	description: "The subscriber's email address.",
	sample: 'priya@example.com',
}

const CONFIRM_URL: EmailVariable = {
	key: 'confirm_url',
	label: 'Confirm link',
	description: 'The link that confirms the subscription. Use it as a button link.',
	sample: 'https://bwh.tech/confirm',
}

const DOWNLOAD_URL: EmailVariable = {
	key: 'download_url',
	label: 'Download link',
	description: 'The link to the lead magnet file. Use it as a button link.',
	sample: 'https://bwh.tech/download',
}

const LEAD_MAGNET: EmailVariable = {
	key: 'lead_magnet',
	label: 'Lead magnet',
	description: 'The title of the lead magnet.',
	sample: 'The Missing Frappe Manual',
}

export const NEWSLETTER_VARIABLES = [FIRST_NAME, EMAIL]
export const CONFIRM_VARIABLES = [FIRST_NAME, EMAIL, CONFIRM_URL]
export const WELCOME_VARIABLES = [FIRST_NAME, EMAIL, DOWNLOAD_URL, LEAD_MAGNET]

// Also matches a link, where the editor URL-encodes the braces and spaces.
const VARIABLE_PATTERN = /(?:\{\{|%7B%7B)(?:\s|%20)*([a-z_]+)(?:\s|%20)*(?:\}\}|%7D%7D)/gi

/** Put the sample values in the HTML, so the preview shows what a reader gets. */
export function fillSamples(html: string, variables: EmailVariable[]): string {
	const samples = new Map(variables.map((variable) => [variable.key, variable.sample]))
	return html.replace(VARIABLE_PATTERN, (match, key: string) => escapeHtml(samples.get(key) ?? match))
}

function escapeHtml(text: string) {
	return text.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;')
}

File: bwh_os/mailing/email_variables.py
"""Variables like {{ first_name }} in list email subjects and content.

The OS editor puts them in as chips. Keep the keys and fallbacks in step with
frontend/src/lib/emailVariables.ts.
"""

import re
from collections.abc import Iterable
from html import escape

import frappe
from frappe import _

# Used when the value is empty
FALLBACKS = {"first_name": "there"}

SUBSCRIBER = ("first_name", "email")
NEWSLETTER = SUBSCRIBER
WELCOME = SUBSCRIBER  # the form's greeting. The file has its own email; see LEAD_MAGNET.
CONFIRM = (*SUBSCRIBER, "confirm_url")
MAGNET = ("download_url", "lead_magnet")
LEAD_MAGNET = (*SUBSCRIBER, *MAGNET)

# Also matches a link, where the editor URL-encodes the braces and spaces.
VARIABLE = re.compile(r"(?:\{\{|%7B%7B)(?:\s|%20)*([a-z_]+)(?:\s|%20)*(?:\}\}|%7D%7D)", re.IGNORECASE)
# Anything in double braces, to catch old Jinja such as {{ first_name or "there" }}.
BRACES = re.compile(r"\{\{(.*?)\}\}")


def fill(text: str, values: dict[str, str | None], html: bool = True) -> str:
	"""Put the values in the text. A variable with no value gets its fallback."""

	def value_for(match: re.Match) -> str:
		key = match.group(1).lower()
		if key not in values:
			return match.group(0)
		value = values[key] or FALLBACKS.get(key, "")
		return escape(value) if html else value

	return VARIABLE.sub(value_for, text or "")


def with_lead_magnet(allowed: Iterable[str], lead_magnet: str | None) -> tuple[str, ...]:
	"""The magnet variables are fillable only when the email carries a magnet."""
	return (*allowed, *MAGNET) if lead_magnet else tuple(allowed)


def fallback_values(allowed: Iterable[str]) -> dict[str, str | None]:
	"""Every variable empty, for a test send or the web archive."""
	return dict.fromkeys(allowed)


def subscriber_values(subscriber) -> dict[str, str | None]:
	return {"first_name": subscriber.first_name, "email": subscriber.email}


def check(text: str, allowed: Iterable[str], label: str, required: Iterable[str] = ()):
	"""Raise for a variable that this email cannot fill, or a required one that is missing."""
	allowed = set(allowed)
	used = {match.group(1).lower() for match in VARIABLE.finditer(text or "")}
	unknown = sorted(used - allowed)
	unknown += [
		f"{{{{{inner}}}}}"
		for inner in BRACES.findall(text or "")
		if not VARIABLE.fullmatch(f"{{{{{inner}}}}}")
	]
	if unknown:
		frappe.throw(
			_("{0} has a variable that it cannot fill: {1}. Use {2}.").format(
				label, ", ".join(unknown), ", ".join(f"{{{{ {key} }}}}" for key in sorted(allowed))
			)
		)
	for key in required:
		if key not in used:
			frappe.throw(_("{0} needs {1}. Add it as a button link.").format(label, f"{{{{ {key} }}}}"))
